src/core/Instrument.cpp

```diff
@@ -78,8 +82,83 @@ bool Instrument::isFromTrack( const Track * _track ) const
 	return( m_instrumentTrack == _track );
 }
 
+int count_zero_crossings(sampleFrame *buf,fpp_t start, fpp_t frames)
+{
+	// zero point crossing counts of all channels
+	int zero_crossings[DEFAULT_CHANNELS] = {0};
+	// maximum zero point crossing of all channels
+	int max_zc = 0;
+
+	// determine the zero point crossing counts
+	for (fpp_t f = start; f < frames; ++f)
+	{
+		for (ch_cnt_t ch=0; ch < DEFAULT_CHANNELS; ++ch)
+		{
+			// we don't want to count [-1, 0, 1] as two crossings
+			if ((buf[f-1][ch] <= 0.0 && buf[f][ch] > 0.0) ||
+					(buf[f-1][ch] >= 0.0 && buf[f][ch] < 0.0))
+			{
+				++zero_crossings[ch];
+				if (zero_crossings[ch] > max_zc)
+				{
+					max_zc = zero_crossings[ch];
+				}
+			}
+		}
+	}
+
+	return max_zc;
+}
+
+
+void Instrument::applyFadeIn(sampleFrame * buf, NotePlayHandle * n)
+{
+	f_cnt_t total = n->totalFramesPlayed();
+	if (total == 0)
+	{
+		const fpp_t frames = n->framesLeftForCurrentPeriod();
+
+		int max_zc = count_zero_crossings(buf, 1, frames);
+
+		// calculate the length of the fade in
+		// Length is inversely proportional to the max of zero_crossings,
+		// because for low frequencies, we need a longer fade in to
+		// prevent clicking.
+		fpp_t length = (fpp_t) (85.0  / ((float)max_zc / ((float)frames/128.0f) + 1.0f));
```

Comment:
I can't follow why you changed the formula. Where are the constants `85` and `128` from? I imagine like this:

* `85.0` is the maximum fade in that you ever want to allow (IIRC we never had more in zyn)
* `128` is used because at 256, we get a perfect fade with zyn's formula, so this was the old `2.0`, and now you're scaling it down to respect the number of frames

Are my assumptions correct?

Reply:
Your assumptions are correct.

The Zyn's formula `((float)frames - 1)  / ((float)max_zc / 2.0f + 1.0f) / 3.0f` calculates the fade-in length pretty well at 256 samples. If we plug `frames=256` into the formula, it becomes `255.0f  / ((float)max_zc / 2.0f + 1.0f) / 3.0f` which simplifies to `85.0f  / ((float)max_zc / 2.0f + 1.0f)`.

This assumes a buffer size of 256 samples, so we need to adapt it to different buffer sizes. Instead of using the number of zero point crossings, we should use the number of zero point crossings per frame. If we divide `max_zc ` by 2 at 256 samples, we need to divide it by 4 at 512 samples, by 1 at 128 samples and so on. This is where 128 comes from.

I think 128 is somewhat understandable, but 85 looks magical and appears out of nowhere. We can define a constant for the max fade in length and replace 85 with that.

Reply:
Do you want to add a further comment for the formula, or should we resolve this comment?

Reply:
Should we also define `128` as a constant? I will make a `const static` variable in function to define `85`, is that all right?

Reply:
Would be OK for me. You can also make it const static in the class under `private` if you want. Probably both is equally good.